SMT-Lib printing: Handle predicate objects from new assumptions

TiloRC opened the pull request:
The smtlib_code function will no longer throw an error when it tries to print some predicate from the new assumptions. The printer throws an error if given a predicate that can’t be translated to SMT-Lib or could be but hasn’t been Implemented. While it’s possible to handle even, odd, and some other predicate, this isn’t really all that important so this functionality was dropped.

Example:

In [1]: from sympy.printing.smtlib import smtlib_code

In [2]: from sympy.assumptions.ask import Q

In [3]:smtlib_code(Q.positive(x))

Out[3]]: '(declare-const x Real)\n(assert (> x 0))'

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

This is a continuation of this pull request #25232. I worked on a lot of stuff that didn't seem to be a priority so this pull request contains the most essential bits of that one without any distracting commits or comments.

#### Other Comments

@kunalsheth if you want to review this it would be appreciated. 

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* printing
  * make SMT-Lib printer handle some predicates from the new assumptions
<!-- END RELEASE NOTES -->

asmeurer commented on sympy/printing/tests/test_smtlib.py:
```diff
@@ -73,6 +74,26 @@ def test_Relational():
+        assert smtlib_code(Q.zero(x), auto_declare=False, log_warn=w) == "(assert (= x 0))"
+        assert smtlib_code(Q.nonpositive(x), auto_declare=False, log_warn=w) == "(assert (<= x 0))"
+        assert smtlib_code(Q.nonnegative(x), auto_declare=False, log_warn=w) == "(assert (>= x 0))"
+        assert smtlib_code(Q.nonzero(x), auto_declare=False, log_warn=w) == "(assert (not (= x 0)))"
```
What happens with unknown predicates? Can we add tests for that?

asmeurer commented:
@kunalsheth I know you wrote the original SMT-LIB printing code in SymPy. You may be interested in some of the changes being made here.

TiloRC commented:
I need to remove the "Started to implement EncodedCNF printing" commit before this ready to merge.

TiloRC commented:
How do I cancel a review request? That happened automatically. I messed up my git rebase and it forgot I merged master which caused it to think there are like 50 more commits than there actually should be.

TiloRC committed:
SMT-Lib printing: Handle predicate objects from new assumptions

The smtlib_code function will no longer throw an error when it tries to print some
predicate from the new assumptions. The printer throws an error if given a
predicate that can’t be translated to SMT-Lib or could be but hasn’t been
Implemented. While it’s possible to handle even, odd, and some other
predicate, this isn’t really all that important so this functionality was dropped.

Example:

In [1]: from sympy.printing.smtlib import smtlib_code

In [2]: from sympy.assumptions.ask import Q

In [3]:smtlib_code(Q.positive(x))

Out[3]]: '(declare-const x Real)\n(assert (> x 0))'

TiloRC commented:
@moorepants you can ignore this pull request.
